src/core/public/application/application_service.tsx

```diff
@@ -17,159 +17,168 @@
+      throw new Error('ApplicationService#setup() must be invoked before start.');
     }
 
-    // Disable registration of new applications
-    this.apps$.complete();
-    this.legacyApps$.complete();
-
-    const legacyMode = injectedMetadata.getLegacyMode();
-    const currentAppId$ = new BehaviorSubject<string | undefined>(undefined);
-    const { availableApps, availableLegacyApps, capabilities } = await this.capabilities.start({
-      apps: new Map([...this.apps$.value].map(([id, { app }]) => [id, app])),
-      legacyApps: this.legacyApps$.value,
-      injectedMetadata,
-    });
-
-    // Only setup history if we're not in legacy mode
-    const history = legacyMode ? null : createBrowserHistory({ basename: http.basePath.get() });
+    this.registrationClosed = true;
+    const { capabilities } = await this.capabilities.start({ injectedMetadata });
+    const availableMounters = filterAvailable(this.mounters, capabilities);
 
     return {
-      availableApps,
-      availableLegacyApps,
+      availableApps: filterAvailable(this.apps, capabilities),
+      availableLegacyApps: filterAvailable(this.legacyApps, capabilities),
       capabilities,
+      currentAppId$: this.currentAppId$.asObservable(),
       registerMountContext: this.mountContext.registerContext,
-      currentAppId$,
-
-      getUrlForApp: (appId, options: { path?: string } = {}) => {
-        return http.basePath.prepend(appPath(appId, options));
-      },
-
+      getUrlForApp: (appId, { path }: { path?: string } = {}) =>
+        getAppUrl(availableMounters, appId, path),
       navigateToApp: (appId, { path, state }: { path?: string; state?: any } = {}) => {
-        if (legacyMode) {
-          // If we're in legacy mode, do a full page refresh to load the NP app.
-          redirectTo(http.basePath.prepend(appPath(appId, { path })));
-        } else {
-          // basePath not needed here because `history` is configured with basename
-          history!.push(appPath(appId, { path }), state);
-        }
-      },
-
-      getComponent: () => {
-        if (legacyMode) {
-          return null;
-        }
-
-        // Filter only available apps and map to just the mount function.
-        const appMounters = new Map<string, AppMounter>(
-          [...this.apps$.value]
-            .filter(([id]) => availableApps.has(id))
-            .map(([id, { mount }]) => [id, mount])
-        );
-
-        return (
-          <AppRouter
-            apps={appMounters}
-            legacyApps={availableLegacyApps}
-            basePath={http.basePath}
-            currentAppId$={currentAppId$}
-            history={history!}
-            redirectTo={redirectTo}
-          />
-        );
+        this.navigate!(getAppUrl(availableMounters, appId, path), state);
+        this.currentAppId$.next(appId);
       },
+      getComponent: () =>
+        this.history ? <AppRouter history={this.history} mounters={availableMounters} /> : null,
```

Comment:
Agreed that it was more explicit, but at the cost of conflating what `AppRouter` really needs. If the logic behind why `AppRouter` needing history changes more than just `isLegacy`, it would lead to less churn by changing when history is generated over sprinkling several `isLegacy && isSomethingElse` checks, or having to rename the variable. Does that make sense? 😺